### PR TITLE
a new module to manage static route and bfd.

### PR DESCRIPTION
**Pull request by yuandongx**
##### SUMMARY

- Manages the static routes and bfd linkage.

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py
##### ADDITIONAL INFORMATION

example:
- name: Config an ip route-static bfd 10GE1/0/1 3.3.3.3 min-rx-interval 50 min-tx-interval 50 detect-multiplier 5
    ce_static_route_bfd:
      function_flag: 'singleBFD'
      nhp_interface: 10GE1/0/1
      next_hop: 3.3.3.3
      min_tx_interval: 50
      min_rx_interval: 50
      detect_multiplier: 5
      aftype: v4
      state: present

dvice will show command:
ip route-static bfd 10GE1/0/1 3.3.3.3 min-rx-interval 50 min-tx-interval 50 detect-multiplier 5

```paste below

```

### CODE REVIEW

**Comment by xuxiaowei0512** [comment]
-label affects_2.9
+label affects_2.10

File: lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py [hunk 1]
```diff
@@ -0,0 +1,1622 @@
...
+    description:
+      - Destination ip address family type of static route.
+    required: true
+    type: str
+    choices: ['v4','v6']
+  next_hop:
+    description:
+      - Next hop address of static route.
+    required: false
+    default: null
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py]: required: false is the default behavior

File: lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py [hunk 2]
```diff
@@ -0,0 +1,1622 @@
...
+    type: str
+  nhp_interface:
+    description:
+      - Next hop interface full name of static route.
+    required: false
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py]: required: false is the default behavior

File: lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py [hunk 3]
```diff
@@ -0,0 +1,1622 @@
...
+    default: null
+    type: str
+  vrf:
+    description:
+      - VPN instance of destination ip address.
+    required: false
+    default: null
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py]: required: false is the default behavior

File: lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py [hunk 4]
```diff
@@ -0,0 +1,1622 @@
...
+  destvrf:
+    description:
+      - VPN instance of next hop ip address.
+    required: false
+    default: null
+    type: str
+  tag:
+    description:
+      - Route tag value (numeric).
+    required: false
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py]: The same

File: lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py [hunk 5]
```diff
@@ -0,0 +1,1622 @@
...
+    default: null
+    type: int
+  description:
+    description:
+      - Name of the route. Used with the name parameter on the CLI.
+    required: false
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py]: same!

File: lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py [hunk 6]
```diff
@@ -0,0 +1,1622 @@
...
+    default: null
+    type: str
+  pref:
+    description:
+      - Preference or administrative difference of route (range 1-255).
+    required: false
+    default: null
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py]: same.

File: lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py [hunk 7]
```diff
@@ -0,0 +1,1622 @@
...
+    description:
+      - Used to distinguish between command line functions.
+    required: true
+    choices: ['globalBFD','singleBFD','dynamicBFD','staticBFD']
+    default: null
+    type: str
+  min_tx_interval:
+    description:
+      - Set the minimum BFD session sending interval (range 50-1000).
+    required: false
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py]: same.

File: lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py [hunk 8]
```diff
@@ -0,0 +1,1622 @@
...
+    description:
+      - Set the minimum BFD receive interval (range 50-1000).
+    required: false
+    default: null
+    type: int
+  detect_multiplier:
+    description:
+      - Configure the BFD multiplier (range 3-50).
+    required: false
+    default: null
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py]: same.

File: lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py [hunk 9]
```diff
@@ -0,0 +1,1622 @@
...
+    description:
+      - bfd name (range 1-15).
+    required: false
+    default: null
+    type: str
+  commands:
+    description:
+      - Incoming command line is used to send sys,undo ip route-static default-bfd,commit.
+    required: false
+    default: null
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py]: same.

File: lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py [hunk 10]
```diff
@@ -0,0 +1,1622 @@
...
+  detect_multiplier:
+    description:
+      - Configure the BFD multiplier (range 3-50).
+    required: false
+    default: null
+    type: int
+  bfd_session_name:
+    description:
+      - bfd name (range 1-15).
+    required: false
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py]: same.

File: lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py [hunk 11]
```diff
@@ -0,0 +1,1622 @@
...
+    returned: always
+    type: dict
+    sample: {"function_flag": "staticBFD", "next_hop": "3.3.3.3", "pref": "100",
+            "prefix": "192.168.20.642", "mask": "24", "description": "testing",
+            "vrf": "_public_", "bfd_session_name": "btoa"}
+existing:
+    description: k/v pairs of existing switchport
+    returned: always
+    type: dict
+    sample: {}
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py]: Miss!!!

File: lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py [hunk 12]
```diff
@@ -0,0 +1,1622 @@
...
+        if self.function_flag == 'globalBFD' and self.state == 'absent':
+            self.commands = self.module.params['commands']
+        # state
+        self.changed = False
+        self.updates_cmd = list()
+        self.results = dict()
+        self.proposed = dict()
+        self.existing = dict()
+        self.end_state = dict()
+
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py:614]: This blank may can be remove.

File: lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py [hunk 13]
```diff
@@ -0,0 +1,1622 @@
...
+                for each_num in addr_list:
+                    if not each_num.isdigit():
+                        return False
+                    if int(each_num) > 255:
+                        return False
+                return True
+            else:
+                if self.prefix.find(':') == -1:
+                    return False
+
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py]: blank can be remove

File: lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py [hunk 14]
```diff
@@ -0,0 +1,1622 @@
...
+                self.static_routes_info["sroute"].append(static_info)
+
+    def _checkparams_(self):
+        """check all input params"""
+        if self.function_flag == 'singleBFD':
+            if not self.next_hop:
+                self.module.fail_json(msg='Error: missing required arguments: next_hop.')
+            if self.state != 'absent':
+                if self.nhp_interface == "Invalid0" and (not self.prefix or self.prefix == '0.0.0.0'):
+                    self.module.fail_json(msg='Error: If an nhp_interface is not configured, '
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py]: If a nhp_interface?

File: lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py [hunk 15]
```diff
@@ -0,0 +1,1622 @@
...
+                                              'the prefix must be configured.')
+
+        if self.function_flag != 'globalBFD':
+            if self.function_flag == 'dynamicBFD' or self.function_flag == 'staticBFD':
+                if not self.mask:
+                    self.module.fail_json(msg='Error: missing required arguments: mask.')
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py]: arguments --> argument

File: lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py [hunk 16]
```diff
@@ -0,0 +1,1622 @@
...
+                # check prefix and mask
+                if not self.mask.isdigit():
+                    self.module.fail_json(msg='Error: Mask is invalid.')
+                # self._convertlentomask_(self.mask)
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py]: Unused code should be removed.

File: lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py [hunk 17]
```diff
@@ -0,0 +1,1622 @@
...
+            if self.function_flag != 'singleBFD' or (self.function_flag == 'singleBFD' and self.destvrf != "_public_"):
+                if not self.prefix:
+                    self.module.fail_json(msg='Error: missing required arguments: prefix.')
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py]: arguments??

File: lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py [hunk 18]
```diff
@@ -0,0 +1,1622 @@
...
+                # convert prefix
+                if not self._convertipprefix_():
+                    self.module.fail_json(msg='Error: The %s is not a valid address' % self.prefix)
+
+            if self.nhp_interface != "Invalid0" and self.destvrf != "_public_":
+                self.module.fail_json(msg='Error: Destination vrf dose no support next hop is interface.')
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py]: does no --> does not

**Review by yuandongxx — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py (hunk 1 above)
Comment (yuandongx) [reply lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py]: Done.

File: lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py (hunk 3 above)
Comment (yuandongx) [reply lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py]: Done.

File: lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py (hunk 5 above)
Comment (yuandongx) [reply lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py]: Done

File: lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py (hunk 14 above)
Comment (yuandongx) [reply lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py]: Done.

File: lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py (hunk 11 above)
Comment (yuandongx) [reply lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py]: Done

File: lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py (hunk 13 above)
Comment (yuandongx) [reply lib/ansible/modules/network/cloudengine/ce_static_route_bfd.py]: Done

**Comment by xuxiaowei0512** [comment]
read_for_review

**Comment by xuxiaowei0512** [comment]
ready_for_review

**Comment by hwDCN** [comment]
shipit

**Review by yuandongxx — APPROVED** [review]

**Comment by xuxiaowei0512** [comment]
bot_status

**Comment by gundalow** [comment]
rebuild_merge

**Comment by xuxiaowei0512** [comment]
@gundalow  Thank you for your greate help.
Have a nice weekend.

**Pull request merged**